### PR TITLE
Server side corfu metrics

### PR DESCRIPTION
**Pull request by PavelZaytsev**
## Overview

Description:

Enable some server metrics.

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java [hunk 1]
```diff
@@ -125,7 +156,20 @@ private void processor() {
                     lastOp = null;
                 } else if (currOp == BatchWriterOperation.SHUTDOWN) {
                     log.warn("Shutting down the write processor");
-                    streamLog.sync(true);
+                    Runnable fsyncRunnable = () -> {
+                        try {
+                            streamLog.sync(true);
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java:163]: let's throw LogUnitException

File: infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java [hunk 2]
```diff
@@ -154,11 +198,24 @@ private void processor() {
...
+                                    writeRecordTimer.get().record(append);
+                                }
+                                else {
+                                    append.run();
+                                }
                                 break;
                             case RANGE_WRITE:
                                 RangeWriteMsg writeRange = (RangeWriteMsg) currOp.getMsg().getPayload();
-                                streamLog.append(writeRange.getEntries());
+                                Runnable appendMultiple = () -> streamLog.append(writeRange.getEntries());
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java:212]: Let's get rid of code duplication by method extraction
```
private void yay(Runnable fsyncRunnable, Optional<Timer> fsyncTimer) {
        if (fsyncTimer.isPresent()) {
            fsyncTimer.get().record(fsyncRunnable);
        } else {
            fsyncRunnable.run();
        }
    }
```

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 3]
```diff
@@ -191,6 +193,9 @@
     // Error code required to detect an ungraceful shutdown.
     private static final int EXIT_ERROR_CODE = 100;
 
+    private static final String DEFAULT_METRICS_LOGGER_NAME = "CorfuMetrics";
+
+    private static final Duration DEFAULT_METRICS_LOGGING_INTERVAL_DURATION = Duration.ofSeconds(1);
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java:198]: don't need DURATION suffix

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java [hunk 4]
```diff
@@ -83,9 +87,12 @@ public SequencerServerCache(int cacheSize, long maxConflictNewSequencer) {
         this.cacheSize = cacheSize;
         conflictKeys = new HashMap();
         cacheEntries = new PriorityQueue(cacheSize, Comparator.comparingLong
-                (a -> ((ConflictTxStream)a).txVersion));
+                (a -> ((ConflictTxStream) a).txVersion));
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java:90]: please rename "a" to something else

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/transferprocessor/ParallelTransferProcessor.java [hunk 5]
```diff
@@ -65,7 +70,14 @@ public ParallelTransferProcessor(StateTransferBatchProcessor stateTransferBatchP
...
 
-        allFutures = CFUtils.allOfOrTerminateExceptionally(allFutures, batchTransferResult);
+        CompletableFuture<Void> future = MicroMeterUtils
+                .timeWhenCompletes(
+                        batchTransferResult,
+                        sample,
+                        "state-transfer.timer", "type", "committed"
+                );
+
+        allFutures = CFUtils.allOfOrTerminateExceptionally(allFutures, future);
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/transferprocessor/ParallelTransferProcessor.java:80]: It seems to me this method is the same as CompletableFuture.allOf(...)

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java [hunk 6]
```diff
@@ -99,6 +103,10 @@ public PollReport poll(
         );
     }
 
+    private void registerTimerForNodeIfNeeded(String node) {
+        MeterRegistryProvider.getInstance().ifPresent(registry -> registry.timer("failure-detector.ping-latency", "node", node));
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java:107]: It seems to me you can get rid of some boilerplate code if you add a "timer" method into MetricRegistryProvider
```
public static void timer(String timerName, String tag, String tag2) {
        getInstance()
                .ifPresent(registry -> registry.timer(timerName, tag, tag2));
    }
```

then you can use it:
```
MeterRegistryProvider.timer("failure-detector.ping-latency", "node", node);
```

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 7]
```diff
@@ -242,6 +254,8 @@ public static CorfuRuntimeParametersBuilder builder() {
          */
         private MetricRegistry metricRegistry;
 
+        private MicroMeterRuntimeConfig microMeterRuntimeConfig;
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:257]: final?

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java [hunk 8]
```diff
@@ -185,57 +205,69 @@ private void validateStateOfWrittenEntry(long address, @Nonnull ILogData ld) {
...
+            } else {
+                ld = new LogData(DataType.DATA, data, runtime.getParameters().getCodecType());
             }
 
-            // Set the data to use the token
-            ld.useToken(token);
-            ld.setId(runtime.getParameters().getClientId());
+            logSizeDist.ifPresent(dist -> dist.record(ld.getSizeEstimate()));
+            layoutHelper(e -> {
+                Layout l = e.getLayout();
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java:219]: please rename "l"

File: test/src/test/java/org/corfudb/integration/AbstractIT.java [hunk 9]
```diff
@@ -511,19 +525,38 @@ public String getOptionsString() {
          */
         public Process runServer() throws IOException {
             final String serverConsoleLogPath = CORFU_LOG_PATH + File.separator + host + "_" + port + "_consolelog";
-
+            System.out.println(serverConsoleLogPath);
```
Comment (xnull) [inline test/src/test/java/org/corfudb/integration/AbstractIT.java:528]: please use logger

File: common/src/main/java/org/corfudb/common/metrics/micrometer/protocoltransformer/influx/TimerInfluxLineTransformer.java [hunk 10]
```diff
@@ -23,6 +23,6 @@
     public boolean test(String s) {
         Matcher matcher = timerPattern.matcher(s);
         this.matched = Optional.of(matcher);
-        return matcher.find();
+        return matcher.matches();
```
Comment (Maithem) [inline common/src/main/java/org/corfudb/common/metrics/micrometer/protocoltransformer/influx/TimerInfluxLineTransformer.java:26]: Can you add tests?

The regex is not correct. The timer outputs a decimal for throughout and not an integer.

The following test will fail
```java

@Test
    public void timerTest() {
        String dataPoint = "rpc-latency{endpoint=localhost:9000} throughput=3.4/s mean=0.000204964s max=0.001044081s";
        TimerInfluxLineTransformer txn = new TimerInfluxLineTransformer();
        assertThat(txn.test(dataPoint)).isTrue();
    }
```

You can fix it with 

```java
protected static final String TIMER_PATTERN_STRING =
            METRIC_PATTERN_STRING + " (throughput=[\\d.]+/s mean=[\\d.]+s max=[\\d.]+s)$";
```

File: infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java [hunk 11]
```diff
@@ -105,7 +123,20 @@ private void processor() {
...
-                        streamLog.sync(sync);
+                        Runnable fsyncRunnable = () -> {
+                            try {
+                                streamLog.sync(sync);
+                            } catch (IOException e) {
+                                throw new RuntimeException(e);
+                            }
+                        };
+                        if (fsyncTimer.isPresent()) {
+                            fsyncTimer.get().record(fsyncRunnable);
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java:134]: This is not the fsync latency, but rather the fsync time of all dirty segments. Based on the current batch + segment size at most it the sum of fsyncing two files.

File: infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java [hunk 12]
```diff
@@ -70,6 +78,16 @@ public BatchProcessor(StreamLog streamLog, long sealEpoch, boolean sync) {
         this.streamLog = streamLog;
         operationsQueue = new LinkedBlockingQueue<>();
         processorService.submit(this::processor);
+
+        writeRecordTimer = MeterRegistryProvider.getInstance().map(registry ->
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java:82]: Having a histogram on the batch size/queue size can help answer questions about queuing delay.

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java (hunk 3 above)
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java:198]: This is a really short period given that most of the metrics collected are aggregates.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java [hunk 13]
```diff
@@ -32,13 +38,21 @@
     //Empirical threshold of number of streams in a logdata beyond which server performance may be slow
     private final int MAX_STREAM_THRESHOLD = 20;
 
+    private final Optional<Timer> readTimer;
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java:41]: Can you also capture stats on the hit ratio, load time and total weight.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 14]
```diff
@@ -155,7 +160,8 @@ public SequencerServer(ServerContext serverContext) {
 
         globalLogTail = Address.getMinAddress();
         this.cache = new SequencerServerCache(config.getCacheSize(), globalLogTail - 1);
-        setUpTimerNameCache();
+        this.txResolutionTimer = MeterRegistryProvider.getInstance().map(registry ->
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:163]: Can you also capture histogram of the number of queues per transaction.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java [hunk 15]
```diff
@@ -81,11 +84,17 @@
...
         maxConflictWildcard = maxConflictNewSequencer;
         this.maxConflictNewSequencer = maxConflictNewSequencer;
+        conflictKeys = MeterRegistryProvider
+                .getInstance()
+                .map(registry ->
+                        registry.gauge(conflictKeysCounterName, Collections.emptyList(),
+                                new HashMap<ConflictTxStream, Long>(), HashMap::size))
+                .orElse(new HashMap<>());
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java:97]: We need to capture stats on the window size and number of evictions per trim call.

File: scripts/corfu_server.sh [hunk 16]
```diff
@@ -35,9 +35,13 @@ then
     CLASSPATH=`cygpath -wp "$CLASSPATH"`
 fi
 
+if [ "$METRICS_CONFIG_FILE" != "" ]; then
+  LOGBACK_CONFIGURATION="-Dlogback.configurationFile=${METRICS_CONFIG_FILE}"
+  JAVA="$JAVA $LOGBACK_CONFIGURATION"
+fi
```
Comment (Maithem) [inline scripts/corfu_server.sh:41]: Can you just enable this by default and add the logging information in infrastructure/src/main/resources/logback.prod.xml

**Review by Maithem — APPROVED** [review]

File: common/src/main/java/org/corfudb/common/metrics/micrometer/MeterRegistryProvider.java [hunk 17]
```diff
@@ -41,6 +62,22 @@ public static synchronized void init(Logger logger, Duration loggingInterval,
...
 
+        /**
+         * Configure the meter registry of type LoggingMeterRegistry. All the metrics registered
+         * with this meter registry will be exported in the InfluxDB line protocol format
+         * (https://docs.influxdata.com/influxdb/v1.8/write_protocols/line_protocol_tutorial/)
+         * with  the provided loggingInterval frequency.
+         * @param logger A configured logger.
+         * @param loggingInterval A duration between log appends for every metric.
+         */
+        public static synchronized void init(Logger logger, Duration loggingInterval) {
```
[the remark this answers is not in the document]
Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/MeterRegistryProvider.java]: Needs to be public to initialize it.

File: common/src/main/java/org/corfudb/common/metrics/micrometer/MicroMeterUtils.java [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+                    " the number of tags should be even.");
+        }
+        Optional<MeterRegistry> maybeRegistry = MeterRegistryProvider.getInstance();
+        if (maybeRegistry.isPresent() && maybeSample.isPresent()) {
+            MeterRegistry meterRegistry = maybeRegistry.get();
+            Timer.Sample sample = maybeSample.get();
+            CompletableFuture<T> cf = new CompletableFuture<>();
+            future.whenComplete((res, ex) -> {
+                sample.stop(meterRegistry.timer(timerName, tags));
+                if (ex != null) {
```
[the remark this answers is not in the document]
Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/MicroMeterUtils.java:29]: Here I need to stop the timer no matter what and return the original result or the exception. I did not find another way of doing it.

File: infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java (hunk 2 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java:193]: Sounds good

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/transferprocessor/ParallelTransferProcessor.java (hunk 5 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/transferprocessor/ParallelTransferProcessor.java:80]: allOf finishes when all the futures finish, here we need to quit once one of the futures terminate exceptionally.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java (hunk 4 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java:97]: done

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java (hunk 3 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java:198]: Done

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java (hunk 6 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java:107]: Done

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java (hunk 7 above)
Comment (PavelZaytsev) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: done

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java (hunk 8 above)
Comment (PavelZaytsev) [reply runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: done

File: test/src/test/java/org/corfudb/integration/AbstractIT.java (hunk 9 above)
Comment (PavelZaytsev) [reply test/src/test/java/org/corfudb/integration/AbstractIT.java]: Was removed

File: common/src/main/java/org/corfudb/common/metrics/micrometer/protocoltransformer/influx/TimerInfluxLineTransformer.java (hunk 10 above)
Comment (PavelZaytsev) [reply common/src/main/java/org/corfudb/common/metrics/micrometer/protocoltransformer/influx/TimerInfluxLineTransformer.java]: Was fixed

File: infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java (hunk 11 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java]: Fixed

File: infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java (hunk 12 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java:83]: Added

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java (hunk 3 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java:198]: This was for debugging. Increased to a minute.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java (hunk 13 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java:44]: Added

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 14 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:169]: Added

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java (hunk 15 above)
Comment (PavelZaytsev) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java:105]: Added

File: scripts/corfu_server.sh (hunk 16 above)
Comment (PavelZaytsev) [reply scripts/corfu_server.sh:41]: This was done in order to supply the running processes in the integration tests with an XML file. It's mostly for debugging/viewing metrics in the integration tests. I will enable the metrics in the logback.prod.xml though.

**Review by hisundar — COMMENTED** [review]
There are multiple implications of the change made here that will increase the amount of logging corfu does today - 
1. change from warn to debug
2. addition of another metrics logging facility

There is already a lot of pressure on corfu to _reduce_ the amount of log produced. This change, while certainly useful, will make it harder for us to justify the increase.
Also I hope that the main reason for disabling metrics in prior releases (CPU and memory hog) has been evaluated and addressed here.

**Review by xnull — APPROVED** [review]
looks good

**Pull request merged**